rhino/src/main/java/org/mozilla/javascript/typedarrays/NativeTypedArrayView.java

```diff
@@ -1236,6 +1244,139 @@ private static Object js_with(Context cx, Scriptable scope, Scriptable thisObj,
+        Scriptable mapFnThisArg = Undefined.SCRIPTABLE_UNDEFINED;
+        if (!Undefined.isUndefined(mapArg)) {
+            if (!(mapArg instanceof Function)) {
+                throw ScriptRuntime.typeErrorById("msg.map.function.not");
+            }
+            mapFn = (Function) mapArg;
+            if (args.length >= 3) {
+                mapFnThisArg = ScriptableObject.ensureScriptable(args[2]);
+            }
+        }
+
+        Object iteratorProp = ScriptableObject.getProperty(items, SymbolKey.ITERATOR);
+        if (!(iteratorProp == Scriptable.NOT_FOUND)
+                // Optimization: Although NativeArrays and TypedArrays have iterators,
+                // we don't want to use them here to avoid copying the contents
+                // to determine the length.
+                // However, with this the test262 test
+                // built-ins/TypedArray/from/iterated-array-changed-by-tonumber.js
+                // doesn't pass.
+                && !(items instanceof NativeArray)
+                && !(items instanceof NativeTypedArrayView<?>)
+                && !Undefined.isUndefined(iteratorProp)) {
+            final Object iterator = ScriptRuntime.callIterator(items, cx, scope);
+            if (!Undefined.isUndefined(iterator)) {
+                return fromIterator(cx, scope, constructable, iterator, mapFn, mapFnThisArg);
+            }
+        }
+
+        return fromArrayLike(cx, scope, constructable, items, mapFn, mapFnThisArg);
+    }
+
+    private static Object fromIterator(
+            Context cx,
+            Scriptable scope,
+            Constructable constructable,
+            Object iterator,
+            Function mapFn,
+            Scriptable mapFnThisArg) {
+
+        List<Object> values = new ArrayList<>();
+        IteratorLikeIterable it = new IteratorLikeIterable(cx, scope, iterator);
+        for (Object temp : it) {
+            values.add(temp);
+        }
+
+        Scriptable result = constructable.construct(cx, scope, new Object[] {values.size()});
+
+        if (!(result instanceof NativeTypedArrayView)) {
+            throw ScriptRuntime.typeErrorById("msg.typed.array.receiver.incompatible", "from");
+        }
+
+        NativeTypedArrayView<?> typedArray = (NativeTypedArrayView<?>) result;
+        for (int k = 0; k < values.size(); k++) {
+            Object temp = values.get(k);
+            if (mapFn != null) {
+                temp = mapFn.call(cx, scope, mapFnThisArg, new Object[] {temp, (long) k});
+            }
+            typedArray.setArrayElement(k, temp);
+        }
+
+        return result;
+    }
+
+    private static Object fromArrayLike(
+            Context cx,
+            Scriptable scope,
+            Constructable constructable,
+            Scriptable items,
+            Function mapFn,
+            Scriptable mapFnThisArg) {
+
+        final int length = (int) AbstractEcmaObjectOperations.lengthOfArrayLike(cx, items);
```

Comment:
Maybe simply casting here is not correct. Maybe you can add a test case that deals with arrays of size > max_int and check if we can support that also.